ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3MultipartUploadPartHandler.java

```diff
@@ -65,13 +107,104 @@ void handle(RestRequest restRequest, RestResponseChannel restResponseChannel,
     // TODO [S3] : set part ttl value
     // restRequest.setArg(TTL, xxx);
 
-    // 3. Intercept the response callback and make needed changes in response headers.
-    Callback<Void> wrappedCallback = (result, exception) -> {
-      if (exception != null) {
-        callback.onCompletion(null, exception);
-        return;
-      }
-      try {
+    restRequest.setArg(SEND_FAILURE_REASON, Boolean.TRUE);
+
+    // TODO [S3] verifyChunkAccountAndContainer? getAndVerifyReservedMetadataBlobId?
+    new S3MultipartUploadPartHandler.CallbackChain(restRequest, restResponseChannel, callback).start();
+  }
+
+  /**
+   * Represents the chain of actions to take. Keeps request context that is relevant to all callback stages.
+   */
+  private class CallbackChain {
+    private final RestRequest restRequest;
+    private final RestResponseChannel restResponseChannel;
+    private final Callback<ReadableStreamChannel> finalCallback;
+    private final String uri;
+
+    /**
+     * @param restRequest the {@link RestRequest}.
+     * @param restResponseChannel the {@link RestResponseChannel}.
+     * @param finalCallback the {@link Callback} to call on completion.
+     */
+    private CallbackChain(RestRequest restRequest, RestResponseChannel restResponseChannel,
+        Callback<ReadableStreamChannel> finalCallback) {
+      this.restRequest = restRequest;
+      this.restResponseChannel = restResponseChannel;
+      this.finalCallback = finalCallback;
+      this.uri = restRequest.getUri();
+    }
+
+    /**
+     * Start the chain by calling {@link SecurityService#processRequest}.
+     */
+    private void start() {
+      restRequest.getMetricsTracker()
+          .injectMetrics(frontendMetrics.putBlobMetricsGroup.getRestRequestMetrics(restRequest.isSslUsed(), false));
+      // Start the callback chain by parsing blob info headers and performing request security processing.
+      securityService.processRequest(restRequest, securityProcessRequestCallback());
+    }
+
+    /**
+     * After {@link SecurityService#processRequest} finishes, call {@link SecurityService#postProcessRequest} to perform
+     * request time security checks that rely on the request being fully parsed and any additional arguments set.
+     * @return a {@link Callback} to be used with {@link SecurityService#processRequest}.
+     */
+    private Callback<Void> securityProcessRequestCallback() {
+      return buildCallback(frontendMetrics.putSecurityProcessRequestMetrics, securityCheckResult -> {
+        //make sure this has been called after processRequest(permission check) since it needs to query dataset db.
```

Comment:
nit: This comment can be removed? Seems to be associated with dataset db

Reply:
Removed. Thanks,